File: programme2_extractor/src/main/java/com/gestion_mp3/Consumer.java
package com.gestion_mp3;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Map;

import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import com.rabbitmq.client.DeliverCallback;

public class Consumer {
    private Connection connexion;
    private Channel channel;
    private static final String HOST = Config.HOST;
    private static final int PORT = Config.PORT;
    private static final String LOGIN = Config.LOGIN;
    private static final String MOT_DE_PASSE = Config.PASSWORD;
    private static final String QUEUE_ENTREE = Config.QUEUE_IN;
    private static final String LOG_FILE_PATH = Config.LOG_FILE_PATH;

    public void initRabbit() throws Exception {

        // 1. Exchange DLX
        channel.exchangeDeclare("mp3_dlx", "direct", true);

        // 2. Queue principale
        channel.queueDeclare(
                QUEUE_ENTREE,
                true,
                false,
                false,
                Map.of(
                        "x-dead-letter-exchange", "mp3_dlx",
                        "x-dead-letter-routing-key", "retry"));

        // 3. Retry queue (attente avant retry)
        channel.queueDeclare(
                "mp3_retry_queue",
                true,
                false,
                false,
                Map.of(
                        "x-dead-letter-exchange", "",
                        "x-dead-letter-routing-key", QUEUE_ENTREE,
                        "x-message-ttl", 5000 // 5 secondes
                ));

        channel.queueBind(
                "mp3_retry_queue",
                "mp3_dlx",
                "retry");

        // 4. Dead Letter Queue finale
        channel.queueDeclare(
                "mp3_dlq",
                true,
                false,
                false,
                null);

        channel.queueBind(
                "mp3_dlq",
                "mp3_dlx",
                "final");

        Logger.writeInLogFile(LOG_FILE_PATH, "[*] RabbitMQ configuré");
    }

    public void connecter() throws Exception {
        ConnectionFactory factory = new ConnectionFactory();
        factory.setHost(HOST);
        factory.setPort(PORT);
        factory.setUsername(LOGIN);
        factory.setPassword(MOT_DE_PASSE);

        connexion = factory.newConnection();
        channel = connexion.createChannel();

        channel.basicQos(1); // Prefetch count = 1

        Logger.writeInLogFile(LOG_FILE_PATH, "[*] En attente de messages sur : " + QUEUE_ENTREE);
    }

    public void consommer(MetaDataExtractor extractor, Producer producer) {

        DeliverCallback callback = (tag, delivery) -> {

            String path = new String(delivery.getBody(), StandardCharsets.UTF_8);

            try {

                Logger.writeInLogFile(LOG_FILE_PATH, "Reçu: " + path);

                Mp3Metadata metadata = extractor.extraire(path);

                producer.publier(metadata);

                channel.basicAck(delivery.getEnvelope().getDeliveryTag(), false);

            } catch (Exception e) {
                try {

                    Map<String, Object> headers = delivery.getProperties().getHeaders();

                    long retryCount = 0;

                    if (headers != null && headers.containsKey("x-retry")) {
                        retryCount = ((Number) headers.get("x-retry")).longValue();
                    }

                    if (retryCount < 3) {

                        AMQP.BasicProperties props = new AMQP.BasicProperties.Builder()
                                .headers(Map.of(
                                        "x-retry",
                                        retryCount + 1))
                                .build();

                        // envoyer dans le DLX avec routing key retry
                        channel.basicPublish(
                                "mp3_dlx",
                                "retry",
                                props,
                                delivery.getBody());

                        // supprimer le message actuel
                        channel.basicAck(
                                delivery.getEnvelope().getDeliveryTag(),
                                false);

                    } else {

                        // envoyer dans la DLQ finale
                        channel.basicPublish(
                                "mp3_dlx",
                                "final",
                                null,
                                delivery.getBody());

                        channel.basicAck(delivery.getEnvelope().getDeliveryTag(), false);
                    }

                } catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        };

        try {
            channel.basicConsume(
                    QUEUE_ENTREE,
                    false,
                    callback,
                    consumerTag -> {
                    });
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void deconnecter() throws Exception {
        if (channel != null)
            channel.close();
        if (connexion != null)
            connexion.close();
    }

}

File: programme3_sender/consumer.js
// consumer.js
const amqp = require('amqplib');
const config = require('./config');
const { writeInLogFile } = require('./logger');

const LOG_FILE_PATH = config.logFilePath;
const MAX_RETRY = 3;

const QUEUE_ENTREE    = config.rabbitmq.queue;       // mp3_metadata
const DLX             = 'mp3_sender_dlx';
const RETRY_QUEUE     = 'mp3_sender_retry_queue';
const DLQ             = 'mp3_sender_dlq';
const TTL             = 5000;   

async function initRabbit(channel) {

    // 1. Exchange DLX
    await channel.assertExchange(DLX, 'direct', { durable: true });

    // 2. Queue principale
    await channel.assertQueue(QUEUE_ENTREE, {
        durable: true,
        arguments: {
            'x-dead-letter-exchange'    : DLX,
            'x-dead-letter-routing-key' : 'retry'
        }
    });

    // 3. Retry queue (attente avant retry)
    await channel.assertQueue(RETRY_QUEUE, {
        durable: true,
        arguments: {
            'x-dead-letter-exchange'    : '',
            'x-dead-letter-routing-key' : QUEUE_ENTREE,
            'x-message-ttl'             : TTL
        }
    });

    await channel.bindQueue(RETRY_QUEUE, DLX, 'retry');

    // 4. Dead Letter Queue finale
    await channel.assertQueue(DLQ, { durable: true });

    await channel.bindQueue(DLQ, DLX, 'final');

    writeInLogFile(LOG_FILE_PATH, '[*] RabbitMQ configuré');
    console.log('[*] RabbitMQ configuré');
}

async function connecter() {
    const { host, port, login, password, queue } = config.rabbitmq;

    const connexion = await amqp.connect(`amqp://${login}:${password}@${host}:${port}`);
    const channel = await connexion.createChannel();

    await channel.assertQueue(queue, { durable: false });
    channel.prefetch(1);

    writeInLogFile(LOG_FILE_PATH,`[*] En attente de messages sur : ${queue}`);

    return { connexion, channel };
}

async function consommer(channel, callback) {
    const { queue } = config.rabbitmq;

    channel.consume(queue, async (message) => {
        if (message === null) return;

        const contenu = message.content.toString();

        try {
            writeInLogFile(LOG_FILE_PATH, `Reçu: ${contenu}`);

            const metadata = JSON.parse(contenu);

            await callback(metadata);

            channel.ack(message);

        } catch (error) {
            try {
                // Récupérer le nombre de tentatives
                const headers = message.properties.headers || {};
                const retryCount = headers['x-retry'] ? Number(headers['x-retry']) : 0;

                if (retryCount < MAX_RETRY) {
                    // Renvoyer dans le DLX avec routing key retry
                    channel.publish(
                        'mp3_dlx',
                        'retry',
                        message.content,
                        {
                            headers: { 'x-retry': retryCount + 1 }
                        }
                    );

                    writeInLogFile(LOG_FILE_PATH, `Retry ${retryCount + 1}/${MAX_RETRY} : ${contenu}`);

                } else {
                    // Envoyer dans la DLQ finale
                    channel.publish(
                        'mp3_dlx',
                        'final',
                        message.content,
                        {}
                    );

                    writeInLogFile(LOG_FILE_PATH, `Echec final, envoyé en DLQ : ${contenu}`);
                }

                // Acquitter le message original dans les deux cas
                channel.ack(message);

            } catch (ex) {
                console.error(`[!] Erreur gestion retry : ${ex.message}`);
            }
        }
    });
}

module.exports = { connecter, consommer };